src/HeapDumpCommon/DotNetHeapInfo.cs

```diff
@@ -136,6 +142,16 @@ void IFastSerializable.FromStream(Deserializer deserializer)
         Gen1End = (Address)deserializer.ReadInt64();
         Gen2End = (Address)deserializer.ReadInt64();
         Gen3End = (Address)deserializer.ReadInt64();
+        // 
+        // TODO: Backward compatibility - changing serialization format should require bumping version and code paths to deal with 
+        // missing values. How?
+        //
+        // In case this value was not saved, there could be two cases:
+        // Case 1: We are running on a runtime without POH, in that case, initializing that to Start works.
+        // Case 2: We are running on a runtime with POH, in that case, POH segments were saved as if
+        //         they were Gen2, there is no way to differentiate, so just pretend they are Gen2.
+        //
+        Gen4End = (Address)deserializer.ReadInt64();
```

Comment:
You won't be able to distinguish between whether the runtime has a POH or not at deserialization time.  If you need to know this at deserialization time, you'll need to add this information and serialize it.  Based on your comment, I think you're fine to not worry about it.

I recommend that you add an implementation of IFastSerializableVersion to GCHeapDumpSegment.  You should set the following values:

Version=1
MinimumVersionCanRead=0
MinimumReaderVersion=1

These numbers mean that the current code can read old versions without Gen4End serialized as well as the current version with Gen4End serialized.  You'll also need to add a check prior to deserializing Gen4End in FromStream to make sure that you only deserialize Gen4End if deserializer.VersionBeingRead >= 1.

Take a look at [https://github.com/microsoft/perfview/blob/main/src/FastSerialization/FastSerialization.cs#L2326](https://github.com/microsoft/perfview/blob/main/src/FastSerialization/FastSerialization.cs#L2326) for more details.